packages/services/src/providers/price/coingecko.provider.ts

```diff
@@ -0,0 +1,432 @@
+        });
+      const response = await withRetry(getTokenInfo, {
+        onRetry: ({ attempt, nextDelayMs, error }) => {
+          this.logger.warn(
+            {
+              error: error instanceof Error ? error.message : String(error),
+              attempt,
+              nextDelayMs,
+              tokenAddress,
+              network,
+            },
+            "Retrying CoinGecko API call",
+          );
+        },
+      });
+      return this.parseOnchainResponse(response);
+    } catch (error) {
+      this.logger.error(
+        {
+          error: error instanceof Error ? error.message : JSON.stringify(error),
+          tokenAddress,
+          network,
+        },
+        `Error fetching price`,
+      );
+      return null;
+    }
+  }
+
+  /**
+   * Gets the current price and market data for a single token
+   * @param tokenAddress - The token contract address
+   * @param chain - The blockchain type (EVM or SVM)
+   * @param specificChain - The specific blockchain identifier
+   * @returns A price report with token information, or null if not found
+   */
+  async getPrice(
+    tokenAddress: string,
+    chain: BlockchainType,
+    specificChain: SpecificChain,
+  ): Promise<PriceReport | null> {
+    try {
+      const network = COINGECKO_NETWORKS[specificChain];
+      if (!network) {
+        this.logger.error(`Unsupported chain: ${specificChain}`);
+        return null;
+      }
+      const address = this.normalizeAddress(tokenAddress);
+      if (this.isBurnAddress(address)) {
+        return {
+          price: 0,
+          symbol: "BURN",
+          token: address,
+          timestamp: new Date(),
+          chain,
+          specificChain,
+          pairCreatedAt: undefined,
+          volume: undefined,
+          liquidity: undefined,
+          fdv: undefined,
+        };
+      }
+      const priceData = await this.fetchPrice(address, network);
+      if (priceData) {
+        return {
+          price: priceData.price,
+          symbol: priceData.symbol,
+          token: address,
+          timestamp: new Date(),
+          chain,
+          specificChain,
+          pairCreatedAt: priceData.pairCreatedAt,
+          volume: priceData.volume,
+          liquidity: priceData.liquidity,
+          fdv: priceData.fdv,
+        };
+      }
+    } catch (error) {
+      this.logger.error(
+        {
+          error: error instanceof Error ? error.message : JSON.stringify(error),
+          tokenAddress,
+          chain,
+          specificChain,
+        },
+        `Error fetching price`,
+      );
+    }
+    return null;
+  }
+
+  /**
+   * Get prices for multiple tokens in batches
+   * @param tokenAddresses - Array of token addresses to fetch prices for
+   * @param chain - Blockchain type (EVM or SVM)
+   * @param specificChain - Specific chain identifier
+   * @returns Map of token addresses to their price information (null if not found)
+   */
+  async getBatchPrices(
```

Comment:
note: i realized there _is_ a "get batch addresses" API from CG. originally, i didnt see this.

i'm in the process of refactoring and including that since iiuc we do use the batch method in the `apps/api`. for now, we just do some promises and fetch on token at a time